### ISSUE TITLE
GGM - BUG - Mirrors don't work properly - Reflections appear upside down and they take several frames to update

### ISSUE THREAD

**Issue by Kravenwolf** [TheGameCreators/GameGuruRepo#4216]
Mirrored surfaces using PBR have an upside-down reflection, and the image they are reflecting does not update in real time.

Video here:
https://youtu.be/2axvJL4pPhA

This issue here #4214 regarding "Auto Exposure" may be related to this apparent delay in the PBR not updating in real time? Something, something, light probes, apparently; according to this post #3565 (thanks @fearlesswee). Sure wish I was a programmer...

**Comment by MonkeyFrogStudio** [comment]
This is actually a light probe issue. As pointed out in your other issues threads (like the auto-exposure one), MAX only has one light probe and it's attached to the player. It only updates every once in a while based on distance moved. As a result, with reflective surfaces especially, you will see the light probe update.

I'd said this in your other thread, so sorry for repeating it here (but repeated for others who may read this thread), other game engines allow you to use and place several light probes to control this sort of thing in areas. But I was told that the Wicked Engine (at least the version MAX is using) only allowed one light probe. As a result, the attached it to the player. Until they find a way to have more than one light probe ... or I should say IF they do this ... this issue will not go away.

**Comment by Kravenwolf** [comment]
**@MonkeyFrogStudio** , Thanks for the info. Someone else pointed that out as well, but I am not familiar with light probes. PBR was relatively new to me, as I had to learn it for MAX, so I was just trying to provide all of the information I could on the matter in case it somehow proves useful to the devteam. Your tutorial saved me a lot of time, btw, so thank you for that :)

I spent quite a bit of time messing around with the issue last night (more than I should have, and it was driving me nuts because I thought it was something I was setting incorrectly), and didn't get anywhere with it. Then I tried replicating the issue with some of the stock Aztec assets and couldn't, so I finally just went to bed.

Anyway, as an artist, I can say that this is a pretty **MAJOR** graphical issue. It **completely** gets in the way of creating "visually stunning" interior spaces, as lighting plays a **HUGE** factor in bringing those visuals up to par, and this is directly interfering with it.

Better visuals == better marketing for MAX. This needs to be addressed. :)

**Comment by MonkeyFrogStudio** [comment]
@Kravenwolf Glad my old tutorial was helpful. I had to learn PBR, too, as most of my previous work was old-school styled graphics (when it came to games).

In most game engines, you have actual light probes as "objects" that you can place in an area and set up how much of the area it influences. So, you don't have just one, but as many as you need and can designate them to influence an entire room or area. That way, when the player walks into the area, it's consistent (no popping as the light probe doesn't move, but is stationary and consistently influencing that area). 

When you have separate light probes like this, you can usually even set up how each one works so that each area can be a bit unique. Unfortunately, MAX, as pointed out, only has the one and, as also pointed out, it's tied to the player. So, it needs to update and that creates the popping and inaccuracies that you've highlighted.

It's good that you're pointing these things out. The more that TGC sees this is an issue (and a pretty big one), the more likely it is (I hope) that perhaps something can be done. Even if Wicked (the current version) doesn't have more than one light probe, perhaps TGC can find a way to add more ... and add them in such a way that we can place them as needed in our levels.

**Comment by bond1mjb** [comment]
@MonkeyFrogStudio @Kravenwolf I pointed this issue out repeatedly during development.  That's why PBR often looks flat out wrong in GGM - because it IS wrong.   At one point someone decided the correct thing to do turn reflectivity completely off for all objects which just killed PBR even further and made things looks terrible.   Luckily that was remedied, but it doesn't fix the underlying problem of light probes not accurately representing the environment.   It gives everything in GGM a disjointed look, which is contrary to what PBR should be doing - which is making models look cohesive and grounded in their environment.

**Comment by MonkeyFrogStudio** [comment]
@bond1mjb Yep. I have an older thread on PBR issues where I believe this, along with other issues, was brought up. Let me see if I can find it because it was quite a discussion, if I recall ...

Here it is:

https://github.com/TheGameCreators/GameGuruRepo/issues/2229

MAX looks good (in my opinion), but something still seems a bit off overall. Metal, for example, almost always looks quite dark compared to in other game engines or in apps like Substance Painter.

**Comment by bond1mjb** [comment]
@MonkeyFrogStudio  MAX CAN look good no doubt, when the PBR stars align and you get a proper light probe and everything is represented as it should.  But when you start to build out a level and have contrasting areas, that's when things start to break down.   The PBR guts are there, but that lack of multiple light probes , or at least light probes that properly capture the environment, can really kill the look.

**Comment by bond1mjb** [comment]
@MonkeyFrogStudio @Kravenwolf That would be great if it could be sorted once and for all.   I know back in the EARLY days when I did the initial Cellar demo for Wicked Engine - it allowed me to manually place light probes and scale them to cover their area of influence, which was really cool and fun.  But maybe that's beyond the scope of the GGM target audience.   But look at these pics fullscreen if you can, it makes a MASSIVE difference in how PBR is rendered.  Everything in the GGM pic looks flat with no surface properties because it has a bad light probe in this particular area.   But look how realistically the Wicked pic is.
![ggmax](https://user-images.githubusercontent.com/87768878/223202646-54175c28-cde7-4b8a-ad2e-6cc0b574ff5d.jpg)
![wicked](https://user-images.githubusercontent.com/87768878/223202675-52d2a77d-f7a4-4c8f-b588-4dfe4f090889.jpg)

**Comment by MonkeyFrogStudio** [comment]
@bond1mjb It better not be "beyond the scope of the GGM target audience". And, to be frank, what you've said sort of upsets me  (not you at all, you'll see what I mean in a moment) because I was TOLD that GGM only has ONE light probe because Wicked did not allow for more than ONE! But you're telling me that Wicked allows for multiple light probes! If this is so, then that definitely needs to be included in MAX! It's not a feature, it's a necessity for the sake of proper PBR!

It also explains, in part, why shiny silver metallic objects tend to look so dark and flat in MAX. 

Sigh.

**Comment by bond1mjb** [comment]
@MonkeyFrogStudio  Yep, you can freely place as many light probes as you wish in Wicked, and scale them to cover the area you need.   It then blends between them in areas of overlap.      I'd settle for an checkbox under "advanced settings" to allow this, because it's in Wicked after all.

In the pic below, the blue boxes represent the light probes (Wicked more accurately calls them environment probes). Look how awesome that copper boiler looks with a proper env probe!

![probes](https://user-images.githubusercontent.com/87768878/223205855-9b64a423-7cac-49cd-81dd-5433708f8f5c.jpg)

**Comment by MonkeyFrogStudio** [comment]
@bond1mjb Sigh ... that's all I have to say. I have no idea why they want to hamper MAX like this.

**Comment by Kravenwolf** [comment]
@bond1mjb, @MonkeyFrogStudio, thank you for all of the additional information. Your before/after cellar demo shots alone should be enough to show anyone why this needs to be addressed. Settling for option 1 is completely unacceptable when option 2 could be a tickbox away from the fingertips of more experienced users under the advanced settings tab!

If PBR is not being used properly, then that's a **MAJOR** problem. If **WICKED ENGINE** is not being used to it's full potentional, that's an even bigger one.

**I do hope that TGC recognizes casual users who just want to drag-and-drop assets to make a game are NOT the only audience/customer base that MAX has.** 

There are most certainly members of the community with high-level skills that are waiting [patiently] for them to get MAX to a point where they can create something that really highlights all of the work they've been putting into the engine for last couple of years. Experienced artists, animators, scripters -- are all a part of this community. 

Don't let us down, and keep up the great work. :)

**Comment by Kravenwolf** [comment]
@LeeBamberTGC , (sorry, I've been away!) Thank you VERY much for adding this feature. It should make a dramatic difference in MAX'S visuals once all of the kinks are worked out. I've noticed @MonkeyFrogStudio seems to be on top of testing the new system, so a big thanks to him, too for that! Can't wait to try them out for myself :)

**I had a few questions I was going to ask on the next stream, but I'll try to ask them here instead (hopefully you'll read them)**

1. Is the original probe(s) that was attached to the player still there or has it been removed?

2. Will the flashlight have the option to have a probe attached to it in the event someone makes a level where the flashlight is the only light source present? Would it need one in this case or would it cause the same issue we had originally?

3. You mentioned in the last broadcast video there's a limit to 8 probes, then the engine will grab whichever one is nearby and basically reuse it? Will WE be able to assign which ones are reused where, or can the engine be trusted with this? I'm wondering what will happen in the case of a multi-story building for example, if the closest probe around for the engine to reuse in one area is one above or below the current floor, which may not be the best choice?

**Comment by MonkeyFrogStudio** [comment]
@Kravenwolf I think I can answer your second one ... at least partially. Even though they are being called "light probes" they don't really have anything to do with lights per se. They are more accurately called either reflection probes or environment probes. What they do is take a sort of snap shot of the area they influence and create a cube map of that area (just like a sky map). This map is then used to create reflections on reflective surfaces and influence areas (like how a metal or shiny plastic would pick up colors from it's surroundings). Because of this, you wouldn't need a probe for your flashlight. Instead, if the flash light is used in an area with a probe, it would just brighten up what is being "reflected" by the probe. 

Basically, these probes are only for areas (like rooms, etc.) that need better representation of lighting for their PBR materials (like how glass windows reflect their environment, how metal ornaments reflect back room colors, etc. 

I hope this helps. And if I've gotten any of this wrong, I'm sure that someone will come behind and correct me.
